### PR TITLE
Unify (Cyl. & Car.) SOR code on GPU

### PR DESCRIPTION
**Pull request by lmh91**

### CODE REVIEW

File: test/benchmarks/SOR_cartesian.jl [hunk 1]
```diff
@@ -154,4 +145,19 @@ calculate_electric_potential!(
     refinement_limits = [0.2, 0.1],
     max_n_iterations = 2000,
     # refinement_limits = missing, 
-)
\ No newline at end of file
+)
+
+if false
```
Comment (fhagemann) [inline test/benchmarks/SOR_cartesian.jl:150]: `if false`?

File: test/benchmarks/SOR_cylindrical.jl [hunk 2]
```diff
@@ -177,20 +148,28 @@ SolidStateDetectors.innerloop!(line_weights, pssrb_cyl, iφ, inφ, ir, inr, rb_t
...
-# apply_boundary_conditions!
 begin
     @btime SolidStateDetectors.apply_boundary_conditions!($pssrb_cyl, $update_even_points, $only2d)
     @btime SolidStateDetectors.apply_boundary_conditions!($pssrb_cyl, $update_uneven_points, $only2d)
 end
 
 
-using CUDAKernels
\ No newline at end of file
+if false
```
Comment (fhagemann) [inline test/benchmarks/SOR_cylindrical.jl:162]: `if false`?

File: src/PotentialCalculation/SuccessiveOverRelaxation/GPU_kernel.jl [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    )
+end
+
+"""
+    function get_sor_kernel(::Type{S}, args...) 
+
+where S is either Cartesian or Cylindrical. 
+
+Developer notes:
+Currently (February 2022), there are some limitation to the `@kernel` macro
```
Comment (fhagemann) [inline src/PotentialCalculation/SuccessiveOverRelaxation/GPU_kernel.jl:136]: Typo: limitations

File: src/PotentialCalculation/SuccessiveOverRelaxation/GPU_kernel.jl [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+Developer notes:
+Currently (February 2022), there are some limitation to the `@kernel` macro 
+of the package KernelAbstractions.jl. Especially, regarding usage of dispatch. 
+
+Thus, we have to write two kernel functions right now for the Cartesian & Cylindrical case:
+`sor_cyl_gpu!` and `sor_car_gpu!`.
+
+Inside kernel functions, everything is (and has to be) inlined and we can make use of multiple dispatch. 
+So in the end we only have to write one function for the kernel, `sor_kernel`, 
+which is than inlined inside the two kernel functions.
```
Comment (fhagemann) [inline src/PotentialCalculation/SuccessiveOverRelaxation/GPU_kernel.jl:144]: Typo: then

File: src/PotentialCalculation/SuccessiveOverRelaxation/GPU_kernel.jl [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+We can also use most of the CPU functions with the restriction that all 
+types have to be independent on the GPU-indices of the kernel. 
+E.g., making use of `i23_is_even_t = Val(iseven(i2 + i3))` and other similar statements is not possible,
+which are used in the CPU implementation for optimization.
+One the GPU (currently) those statements have to be calculated and booleans have to be passed.
```
Comment (fhagemann) [inline src/PotentialCalculation/SuccessiveOverRelaxation/GPU_kernel.jl:150]: Type: On

**Review by fhagemann — COMMENTED** [review]
Looks good, I will probably have a deeper look into this in the future.
I have some things I spotted while having a glimpse at your changes

**Comment by fhagemann** [comment]
Also, now that it is `PotentialCalculationSetup` instead of `PotentialSimulationSetup`, the naming `pssrb` does not really make sense anymore.

**Comment by lmh91** [comment]
Thanks for the comments! I will merge as soon as the tests are through. 
Next step will be adding convergence check for the GPU and actually unify that with the CPU implementation of the convergence check. I already have something in the making.

**Pull request merged**